python/src/iceberg/catalog/base.py

```diff
@@ -0,0 +1,251 @@
+        """Create a namespace in the catalog.
+
+        Args:
+            namespace: The namespace to be created. Tuple of namespace levels. Ex: ('com','org','dept')
+            properties: A string dictionary of properties for the given namespace
+
+        Raises:
+            AlreadyExistsError: If a namespace with the name already exists in the namespace
+        """
+
+    @abstractmethod
+    def drop_namespace(self, namespace: Tuple[str, ...]) -> None:
+        """Drop a namespace.
+
+        Args:
+            namespace: The namespace to be dropped. Tuple of namespace levels. Ex: ('com','org','dept')
+
+        Raises:
+            NamespaceNotFoundError: If a namespace with the name does not exist in the namespace
+            NamespaceNotEmptyError: If the namespace is not empty
+        """
+
+    @abstractmethod
+    def list_tables(self, namespace: Optional[Tuple[str, ...]] = None) -> List[Tuple[Tuple[str, ...], str]]:
+        """List tables under the given namespace in the catalog.
+
+        If namespace not provided, will list all tables in the catalog.
+
+        Args:
+            namespace: the namespace to search. Tuple of namespace levels. Ex: ('com','org','dept')
+
+        Returns:
+            List[Tuple[str, str]]: list of tuple of table namespace and their names.
+
+        Raises:
+            NamespaceNotFoundError: If a namespace with the name does not exist in the namespace
+        """
+
+    @abstractmethod
+    def list_namespaces(self) -> List[Tuple[str, ...]]:
+        """List namespaces from the given namespace. If not given, list top-level namespaces from the catalog.
+
+        Returns:
+            List[Tuple[str, ...]]: a List of namespace, where each element is a Tuple of namespace levels. Ex: ('com','org','dept')
+        """
+
+    @abstractmethod
+    def load_namespace_metadata(self, namespace: Tuple[str, ...]) -> Dict[str, str]:
+        """Get metadata dictionary for a namespace.
+
+        Args:
+            namespace: A Tuple of namespace levels. Ex: ('com','org','dept')
+
+        Returns:
+            Dict[str, str]: a dictionary of properties for the given namespace
+
+        Raises:
+            NamespaceNotFoundError: If a namespace with the name does not exist in the namespace
+        """
+
+    @abstractmethod
+    def set_namespace_metadata(self, namespace: Tuple[str, ...], metadata: Dict[str, str]) -> None:
+        """Update or remove metadata for a namespace.
+
+        Note: Existing metadata is overridden, use get, mutate, and then set.
+
+        Args:
+            namespace: A Tuple of namespace levels. Ex: ('com','org','dept')
+            metadata: a dictionary of properties for the given namespace
+
+        Raises:
+            NamespaceNotFoundError: If a namespace with the name does not exist in the namespace
+        """
+
+
+class TableNotFoundError(Exception):
```

Comment:
This is the first time we'll be adding custom exceptions. Since these will probably be used in other parts of the library, I think we should create a file `src/iceberg/exceptions.py` at the root and define these there.

Also, (mostly opinion here), I'd rather not override `__init__` just to provide a default message. I've actually found that default messages make debugging harder and I'd rather we simply define the exception types and raise more explicit messages based on the context where they're raise.

_exceptions.py_
```py
class TableNotFoundError(Exception):
    """Raised when a referenced table cannot be found"""
    pass
```
Then we can require providing a more detailed message whenever it's raised:

_catalog/base.py_
```py
raise TableNotFoundError(f"Cannot load table {table}, not found in catalog {catalog}")
```
```py
raise TableNotFoundError(f"Cannot drop table {table}, not found in catalog {catalog}")
```

Reply:
Done. Removed the custom messaging for errors and moved them over to `src/iceberg/exceptions.py` as suggested